### ISSUE TITLE
Private attributes set in GenerateRules are not available in Imports

### ISSUE THREAD

**Issue by HALtheWise** [bazelbuild/bazel-gazelle#923]
### What version of gazelle are you using?

Latest `master`

### What operating system and processor architecture are you using?

Linux x86_64

### What did you do?

When SetPrivateAttr(...) is used in GenerateRules, and Imports() is later called with that rule, the copy of the rule provided to Imports does not contain the private attribute. Notably, when Resolve() is later called, it _does_ get a copy with the attribute.

Run `bazel run //:gazelle` with the patch in https://github.com/HALtheWise/bazel-gazelle/commit/c216a0ad028800d82054e4b0047994f82c53356e for a minimal repro

### What did you expect to see?

```
Setting PrivateAttr in language/proto/gen
Imports called on language/proto/gen
         PrivateAttr was SUCCESS
Setting PrivateAttr in language/proto
Imports called on language/proto
         PrivateAttr was SUCCESS
...
Resolve called on @bazel_gazelle//internal/language:all_files
         PrivateAttr was SUCCESS
Resolve called on @bazel_gazelle//internal/version:all_files
         PrivateAttr was SUCCESS
```

### What did you see instead?

```
Setting PrivateAttr in language/proto/gen
Imports called on language/proto/gen
         PrivateAttr was <nil>
Setting PrivateAttr in language/proto
Imports called on language/proto
         PrivateAttr was <nil>
...
Resolve called on @bazel_gazelle//internal/language:all_files
         PrivateAttr was SUCCESS
Resolve called on @bazel_gazelle//internal/version:all_files
         PrivateAttr was SUCCESS
```

**Comment by jayconrod** [comment]
Ah, `Imports` makes more sense. `Imports` is called the merged rules in the build file. `Resolve` is called on the rules originally generated by `GenerateRules`. So that's why `Imports` doesn't see the private attributes.

I think it's important for `Imports` not to rely on private attributes though. If Gazelle is indexing a directory but not updating it (which happens when you run Gazelle with positional arguments or with `-r=false`), then it will call `Imports` on the rules in the build file (which can't have private attributes).

Could you say more about the extension you're building? I can think of a few ways around this, but I'm not sure whether they're appropriate for your situation.

* If there's a strong convention that can be followed based on the repository name and directory path, just use that in `Resolve` instead of returning anything in `Imports`. Most rules work like this within Google.
* Add an explicit attribute listing import strings in `GenerateRules`, then set it to an empty list in `Resolve`, which will cause `merger` to delete it. To make sure it's merged in both passes, add the attribute to the `MergeableAttrs` and `ResolveAttrs` sets for the rule's `KindInfo`. The Go extension worked this way before private attributes were introduced.
* Add an explicit attribute listing import strings to the rule. Gazelle could generate and use that instead of a private attribute.

**Comment by HALtheWise** [comment]
That makes sense, I'm OK with using one of those workarounds.

The extension I'm working on is for an internal Protobuf-like code generator, where the current import structure is both implicit and not repo-relative. Some day, we definitely want to change that to use absolute imports and follow suggestion 1, but I really don't want to block the transition to Bazel+Gazelle until that happens. Effectively, each file defines a (potentially very large) set of types, which then can be referenced from any other file using just their globally unique name. Yes, it's kinda a mess. The lists are long enough and fluctuate frequently enough that I'd rather not use option 3, although I did try it in the past.

I previously was using basically approach 2, and will probably go back to that. One advantage which I wasn't using before is the ability to manually specify elements of "_gazelle_importable_as" in the BUILD file, to cover cases where Gazelle can't automatically parse it from the file structure.

I've been trying hard to get this right, since I'm hoping to soon make plugins for Python and C++, which have significantly more complicated import structures.

**Comment by HALtheWise** [comment]
To follow up on this, I ended up with a solution where there are two "normal" attrs that I set on the rule in GenerateRules, `_gazelle_importable_as` and `_gazelle_no_auto_importable`. The first is set to a list of import strings, and the second is set to True. In Imports, the set of ImportSpecs is set as everything in the first list, plus (if the second is not set to True), the set of importspecs found by scanning the source files. Both attrs are deleted in Resolve.

That also allows the nice behavior of manually specifying additional import strings in the BUILD file with a #keep comment, and they will get handled as if they were from a cached file scanning operation.
